File: apps/api/gridpassport/forecast.py
"""Deterministic forecaster — mirrors packages/core/src/forecast.ts."""

from __future__ import annotations

from datetime import datetime

from .policy import POLICY_VERSION
from .schemas import (
    CaseInput,
    DerivedProof,
    FlexibilityPassport,
    FlexResponseClass,
    PrivateProfile,
    PublicEvidence,
    ReadinessClass,
    RequestRecord,
    RiskClass,
    ScenarioOverride,
)

_RISK_PENALTY: dict[RiskClass, int] = {"low": 0, "medium": 8, "high": 18}
_FLOOD_PENALTY: dict[RiskClass, int] = {"low": 0, "medium": 3, "high": 12}


def _clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


def _firmness_score(p: PrivateProfile, e: PublicEvidence, requested_mw: float) -> int:
    # Derivation-transparency part 2: quantize to nearest 5 to blur the
    # linear-combination inversion. Mirror of packages/core/src/forecast.ts.
    site_bonus = 10 if e.siteControlEvidence else 0
    bess_share = _clamp((p.bessMW / requested_mw) * 50, 0, 10)
    redundancy = _clamp(p.redundancyShiftPercent * 0.4, 0, 10)
    confidence = p.internalScheduleConfidence * 35
    permit_penalty = _RISK_PENALTY[e.permitRisk]
    flood_penalty = _FLOOD_PENALTY[e.floodRisk]
    raw = 25 + confidence + redundancy + bess_share + site_bonus - permit_penalty - flood_penalty
    clamped = _clamp(raw, 0, 100)
    return int(round(clamped / 5) * 5)


def _expected_peak(p: PrivateProfile, requested_mw: float) -> tuple[int, int]:
    # Derivation-transparency part 2: confidence-class-keyed tier bands.
    # Three tiers each map a confidence range to a fixed band, so the
    # exact private confidence is not observer-invertible. Mirror of TS.
    conf = p.internalScheduleConfidence
    if conf >= 0.75:
        low_frac, high_frac = 0.60, 0.80
    elif conf >= 0.55:
        low_frac, high_frac = 0.50, 0.75
    else:
        low_frac, high_frac = 0.40, 0.65
    return (int(round(requested_mw * low_frac)), int(round(requested_mw * high_frac)))


def _response_class(flex_percent: float) -> FlexResponseClass:
    if flex_percent >= 20:
        return "B"
    return "C"


def _duration_band(bess_hours: float) -> tuple[int, int]:
    # Mirror of packages/core/src/forecast.ts::durationBand. A band (not the
    # private bessHours) hides the exact input — [4, 8] matches
    # bessHours ∈ {4, 5, 6, 7}.
    if bess_hours < 4:
        return (2, 4)
    if bess_hours < 8:
        return (4, 8)
    return (8, 12)


def _flex_band(flex_percent: float, requested_mw: float) -> tuple[int, int]:
    # Derivation-transparency part 2: class-keyed tier bands. Observer
    # narrows flex_percent to a tier (3-tier resolution), not the exact
    # value. Mirror of packages/core/src/forecast.ts::flexibilityBand.
    if flex_percent >= 20:
        return (int(round(requested_mw * 0.20)), int(round(requested_mw * 0.40)))
    if flex_percent >= 10:
        return (int(round(requested_mw * 0.10)), int(round(requested_mw * 0.20)))
    return (0, int(round(requested_mw * 0.10)))


def _flex_passport(p: PrivateProfile, requested_mw: float) -> FlexibilityPassport:
    mw_min, mw_max = _flex_band(p.flexPercent, requested_mw)
    d_min, d_max = _duration_band(p.bessHours)
    return FlexibilityPassport(
        mwMin=mw_min,
        mwMax=mw_max,
        durationHoursMin=d_min,
        durationHoursMax=d_max,
        responseClass=_response_class(p.flexPercent),
    )


def _site_readiness(e: PublicEvidence) -> ReadinessClass:
    if not e.siteControlEvidence or e.permitRisk == "high" or e.zoningRisk == "high":
        return "red"
    if e.permitRisk == "medium" or e.floodRisk == "medium" or e.zoningRisk == "medium":
        return "yellow"
    return "green"


def _energization_band(req: CaseInput, readiness: ReadinessClass, flex_percent: float) -> str:
    cod = datetime.fromisoformat(req.targetCOD)
    base_year = cod.year
    base_quarter = (cod.month - 1) // 3 + 1
    flex_shift = -1 if flex_percent >= 20 else (0 if flex_percent >= 10 else 1)
    spread = {"green": 1, "yellow": 2, "red": 4}[readiness]
    start_abs = base_quarter + flex_shift
    end_abs = start_abs + spread

    def fmt(abs_q: int) -> str:
        y = base_year + (abs_q - 1) // 4
        q = ((abs_q - 1) % 4 + 4) % 4 + 1
        return f"Q{q} {y}"

    return f"{fmt(start_abs)} – {fmt(end_abs)}"


def _cost_exposure(firmness: int, permit_risk: RiskClass) -> RiskClass:
    adjusted = firmness - (15 if permit_risk == "high" else 6 if permit_risk == "medium" else 0)
    if adjusted >= 70:
        return "low"
    if adjusted >= 45:
        return "medium"
    return "high"


def _top_blockers(req: CaseInput, readiness: ReadinessClass, firmness: int) -> list[str]:
    blockers: list[str] = []
    p, e = req.privateProfile, req.publicEvidence
    if e.permitRisk == "high":
        blockers.append("Air-permit review tier likely requires full modeling pass")
    elif e.permitRisk == "medium":
        blockers.append("Generator fleet permitting complexity")
    if not e.siteControlEvidence:
        blockers.append("Site control evidence not yet on file")
    if p.internalScheduleConfidence < 0.6:
        blockers.append("Applicant schedule confidence below planning threshold")
    if e.floodRisk != "low":
        label = "High" if e.floodRisk == "high" else "Moderate"
        blockers.append(f"{label} flood overlay on parcel envelope")
    if readiness == "yellow":
        blockers.append(
            "Site plan maturity below threshold for fast-track interconnection study"
        )
    if firmness < 50:
        blockers.append("Secondary transformer lead-time uncertainty")
    return blockers[:3]


def forecast(input_: CaseInput, override: ScenarioOverride | None = None) -> DerivedProof:
    effective = input_.model_copy(deep=True)
    if override and override.flexPercent is not None:
        effective.privateProfile.flexPercent = override.flexPercent
    p, e = effective.privateProfile, effective.publicEvidence
    firmness = _firmness_score(p, e, effective.requestedMW)
    readiness = _site_readiness(e)
    return DerivedProof(
        firmnessScore=firmness,
        expectedPeakMW=_expected_peak(p, effective.requestedMW),
        flexibilityPassport=_flex_passport(p, effective.requestedMW),
        siteReadinessClass=readiness,
        energizationBand=_energization_band(effective, readiness, p.flexPercent),
        costExposureClass=_cost_exposure(firmness, e.permitRisk),
        topBlockers=_top_blockers(effective, readiness, firmness),
        generatedFromPolicyVersion=POLICY_VERSION,
    )


def build_record(input_: CaseInput, override: ScenarioOverride | None = None) -> RequestRecord:
    effective = input_.model_copy(deep=True)
    if override and override.flexPercent is not None:
        effective.privateProfile.flexPercent = override.flexPercent
    proof = forecast(input_, override)
    return RequestRecord(**effective.model_dump(), derivedProof=proof)

File: packages/core/src/forecast.ts
import { POLICY_VERSION } from "./policy";
import type {
  CaseInput,
  DerivedProof,
  FlexResponseClass,
  FlexibilityPassport,
  PrivateProfile,
  PublicEvidence,
  ReadinessClass,
  RequestRecord,
  RiskClass,
  ScenarioOverride,
} from "./types";

const clamp = (v: number, lo: number, hi: number) =>
  Math.max(lo, Math.min(hi, v));

const RISK_PENALTY: Record<RiskClass, number> = { low: 0, medium: 8, high: 18 };
const FLOOD_PENALTY: Record<RiskClass, number> = { low: 0, medium: 3, high: 12 };

// Quantize to nearest 5 to blur the linear-combination inversion (an
// observer with the formula could otherwise recover a weighted sum of
// confidenceContribution + redundancyContribution + bessShare; rounding
// to 5 makes the residual ambiguity basin wider). Derivation-transparency
// part 2 follow-on; pairs with expectedPeakBand + flexibilityPassport
// tier-banding below. Mirror in apps/api/gridpassport/forecast.py.
function firmnessScore(p: PrivateProfile, e: PublicEvidence, requestedMW: number): number {
  const siteBonus = e.siteControlEvidence ? 10 : 0;
  const bessShare = clamp((p.bessMW / requestedMW) * 50, 0, 10);
  const redundancyContribution = clamp(p.redundancyShiftPercent * 0.4, 0, 10);
  const confidenceContribution = p.internalScheduleConfidence * 35;
  const permitPenalty = RISK_PENALTY[e.permitRisk];
  const floodPenalty = FLOOD_PENALTY[e.floodRisk];
  const raw =
    25 +
    confidenceContribution +
    redundancyContribution +
    bessShare +
    siteBonus -
    permitPenalty -
    floodPenalty;
  // Coarse quantization: nearest 5. 73.4 → 75; 62.2 → 60.
  const quantized = Math.round(clamp(raw, 0, 100) / 5) * 5;
  return quantized;
}

// Coarse peak-load tier bands keyed on confidence class, not on the
// raw confidence float. Three tiers (low / medium / high) each map a
// confidence range to a band, so confidence ∈ [0.75, 1.0] all publish
// the same [0.60, 0.80] band — observer can't invert to the exact
// private confidence. Mirror in apps/api/gridpassport/forecast.py.
function expectedPeakBand(
  p: PrivateProfile,
  requestedMW: number,
): [number, number] {
  const conf = p.internalScheduleConfidence;
  let lowFrac: number;
  let highFrac: number;
  if (conf >= 0.75) {
    lowFrac = 0.60;
    highFrac = 0.80;
  } else if (conf >= 0.55) {
    lowFrac = 0.50;
    highFrac = 0.75;
  } else {
    lowFrac = 0.40;
    highFrac = 0.65;
  }
  return [Math.round(requestedMW * lowFrac), Math.round(requestedMW * highFrac)];
}

function responseClass(flexPercent: number): FlexResponseClass {
  if (flexPercent >= 20) return "B";
  if (flexPercent >= 10) return "C";
  return "C";
}

// Coarse commercial BESS duration tiers. Publishing a band means an
// observer can't invert to the exact private bessHours — [4, 8] matches
// bessHours ∈ {4, 5, 6, 7}. Mirror in apps/api/gridpassport/forecast.py.
function durationBand(bessHours: number): [number, number] {
  if (bessHours < 4) return [2, 4];
  if (bessHours < 8) return [4, 8];
  return [8, 12];
}

// Coarse flexibility-MW tier bands keyed on the same class boundaries as
// responseClass. Class-B (flex ≥ 20%) publishes [20% × req, 40% × req];
// two Class-C sub-tiers split at flex=10%. Observer can narrow flexPercent
// to a tier (3-tier resolution), not to the exact value. Mirror in
// apps/api/gridpassport/forecast.py.
function flexibilityBand(
  flexPercent: number,
  requestedMW: number,
): [number, number] {
  if (flexPercent >= 20) {
    return [Math.round(requestedMW * 0.20), Math.round(requestedMW * 0.40)];
  }
  if (flexPercent >= 10) {
    return [Math.round(requestedMW * 0.10), Math.round(requestedMW * 0.20)];
  }
  return [0, Math.round(requestedMW * 0.10)];
}

function flexibilityPassport(
  p: PrivateProfile,
  requestedMW: number,
): FlexibilityPassport {
  const [mwMin, mwMax] = flexibilityBand(p.flexPercent, requestedMW);
  const [durationHoursMin, durationHoursMax] = durationBand(p.bessHours);
  return {
    mwMin,
    mwMax,
    durationHoursMin,
    durationHoursMax,
    responseClass: responseClass(p.flexPercent),
  };
}

function siteReadiness(e: PublicEvidence): ReadinessClass {
  if (
    !e.siteControlEvidence ||
    e.permitRisk === "high" ||
    e.zoningRisk === "high"
  ) {
    return "red";
  }
  if (
    e.permitRisk === "medium" ||
    e.floodRisk === "medium" ||
    e.zoningRisk === "medium"
  ) {
    return "yellow";
  }
  return "green";
}

function energizationBand(
  req: CaseInput,
  readiness: ReadinessClass,
  flexPercent: number,
): string {
  const codDate = new Date(req.targetCOD);
  const baseYear = codDate.getUTCFullYear();
  const baseQuarter = Math.floor(codDate.getUTCMonth() / 3) + 1;
  const flexShift = flexPercent >= 20 ? -1 : flexPercent >= 10 ? 0 : 1;
  const readinessSpread = { green: 1, yellow: 2, red: 4 }[readiness];
  const startAbs = baseQuarter + flexShift;
  const endAbs = startAbs + readinessSpread;
  const fmt = (abs: number): string => {
    const y = baseYear + Math.floor((abs - 1) / 4);
    const q = ((((abs - 1) % 4) + 4) % 4) + 1;
    return `Q${q} ${y}`;
  };
  return `${fmt(startAbs)} – ${fmt(endAbs)}`;
}

function costExposureClass(firmness: number, permitRisk: RiskClass): RiskClass {
  const adjusted = firmness - (permitRisk === "high" ? 15 : permitRisk === "medium" ? 6 : 0);
  if (adjusted >= 70) return "low";
  if (adjusted >= 45) return "medium";
  return "high";
}

function topBlockers(
  req: CaseInput,
  readiness: ReadinessClass,
  firmness: number,
): string[] {
  const blockers: string[] = [];
  const { privateProfile: p, publicEvidence: e } = req;
  if (e.permitRisk === "high") {
    blockers.push("Air-permit review tier likely requires full modeling pass");
  } else if (e.permitRisk === "medium") {
    blockers.push("Generator fleet permitting complexity");
  }
  if (!e.siteControlEvidence) {
    blockers.push("Site control evidence not yet on file");
  }
  if (p.internalScheduleConfidence < 0.6) {
    blockers.push("Applicant schedule confidence below planning threshold");
  }
  if (e.floodRisk !== "low") {
    blockers.push(`${e.floodRisk === "high" ? "High" : "Moderate"} flood overlay on parcel envelope`);
  }
  if (readiness === "yellow") {
    blockers.push("Site plan maturity below threshold for fast-track interconnection study");
  }
  if (firmness < 50) {
    blockers.push("Secondary transformer lead-time uncertainty");
  }
  return blockers.slice(0, 3);
}

export function forecast(
  input: CaseInput,
  override?: ScenarioOverride,
): DerivedProof {
  const effectiveProfile: PrivateProfile = {
    ...input.privateProfile,
    flexPercent:
      override?.flexPercent ?? input.privateProfile.flexPercent,
  };
  const adjusted: CaseInput = {
    ...input,
    privateProfile: effectiveProfile,
  };
  const firmness = firmnessScore(
    adjusted.privateProfile,
    adjusted.publicEvidence,
    adjusted.requestedMW,
  );
  const readiness = siteReadiness(adjusted.publicEvidence);
  return {
    firmnessScore: firmness,
    expectedPeakMW: expectedPeakBand(
      adjusted.privateProfile,
      adjusted.requestedMW,
    ),
    flexibilityPassport: flexibilityPassport(
      adjusted.privateProfile,
      adjusted.requestedMW,
    ),
    siteReadinessClass: readiness,
    energizationBand: energizationBand(
      adjusted,
      readiness,
      adjusted.privateProfile.flexPercent,
    ),
    costExposureClass: costExposureClass(
      firmness,
      adjusted.publicEvidence.permitRisk,
    ),
    topBlockers: topBlockers(adjusted, readiness, firmness),
    generatedFromPolicyVersion: POLICY_VERSION,
  };
}

export function buildRecord(
  input: CaseInput,
  override?: ScenarioOverride,
): RequestRecord {
  const effectiveInput: CaseInput = override?.flexPercent !== undefined
    ? {
        ...input,
        privateProfile: {
          ...input.privateProfile,
          flexPercent: override.flexPercent,
        },
      }
    : input;
  return {
    ...effectiveInput,
    derivedProof: forecast(input, override),
  };
}
